diff --git a/fs/notify/fdinfo.c b/fs/notify/fdinfo.c
--- a/fs/notify/fdinfo.c
+++ b/fs/notify/fdinfo.c
@@ -106,16 +106,71 @@ static void inotify_fdinfo(struct seq_file *m, struct fsnotify_mark *mark)
 	inode_mark = container_of(mark, struct inotify_inode_mark, fsn_mark);
 	inode = igrab(mark->connector->inode);
 	if (inode) {
+#ifdef CONFIG_KSU_SUSFS_SUS_KSTAT
+		if (susfs_is_current_app_uid()) {
+			bool is_fuse = false;
+			if (susfs_is_inode_sus_kstat(inode, &is_fuse)) {
+				unsigned long ino = inode->i_ino;
+				dev_t dev = inode->i_sb->s_dev;
+				susfs_sus_kstat_spoof_inotify_fdinfo(&ino, &dev);
+				seq_printf(m, "inotify wd:%x ino:%lx sdev:%x mask:%x ignored_mask:0 ",
+						inode_mark->wd, ino, dev,
+						(mark->mask & IN_ALL_EVENTS));
+				show_mark_fhandle(m, inode);
+				seq_putc(m, '\n');
+				iput(inode);
+				return;
+			}
+		}
+#endif // #ifdef CONFIG_KSU_SUSFS_SUS_KSTAT
+#ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
+		if (likely(susfs_is_current_proc_umounted())) {
+			struct mount *mnt = real_mount(file->f_path.mnt);
+			if (mnt->mnt_id >= DEFAULT_KSU_MNT_ID) {
+				struct path path;
+				char *pathname = kmalloc(PAGE_SIZE, GFP_KERNEL);
+				char *dpath;
+				if (!pathname) {
+					goto orig_flow;
+				}
+				dpath = d_path(&file->f_path, pathname, PAGE_SIZE);
+				if (!dpath) {
+					goto out_kfree;
+				}
+				if (kern_path(dpath, 0, &path)) {
+					goto out_kfree;
+				}
+				if (!d_backing_inode(path.dentry)) {
+					goto out_path_put;
+				}
+				seq_printf(m, "inotify wd:%x ino:%lx sdev:%x mask:%x ignored_mask:0 ",
+						inode_mark->wd, d_backing_inode(path.dentry)->i_ino, d_backing_inode(path.dentry)->i_sb->s_dev,
+						(mark->mask & IN_ALL_EVENTS));
+				show_mark_fhandle(m, d_backing_inode(path.dentry));
+				seq_putc(m, '\n');
+				path_put(&path);
+				kfree(pathname);
+				iput(inode);
+				return;
+out_path_put:
+				path_put(&path);
+out_kfree:
+				kfree(pathname);
+			}
+		}
+orig_flow:
+		; /* empty statement for gnu89 compat: label must be followed by a statement */
+#endif // #ifdef CONFIG_KSU_SUSFS_SUS_MOUNT
 		/*
 		 * IN_ALL_EVENTS represents all of the mask bits
 		 * that we expose to userspace.  There is at
 		 * least one bit (FS_EVENT_ON_CHILD) which is
 		 * used only internally to the kernel.
 		 */
 		u32 mask = mark->mask & IN_ALL_EVENTS;
-		seq_printf(m, "inotify wd:%x ino:%lx sdev:%x mask:%x ignored_mask:%x ",
-			   inode_mark->wd, inode->i_ino, inode->i_sb->s_dev,
-			   mask, mark->ignored_mask);
+		seq_printf(m, "inotify wd:%x ino:%lx sdev:%x mask:%x ignored_mask:0 ",
+			   inode_mark->wd, inode->i_ino, inode->i_sb->s_dev,
+			   mask);
 		show_mark_fhandle(m, inode);
 		seq_putc(m, '\n');
 		iput(inode);
diff --git a/fs/proc/task_mmu.c b/fs/proc/task_mmu.c
--- a/fs/proc/task_mmu.c
+++ b/fs/proc/task_mmu.c
@@ -22,6 +22,10 @@
 #include <linux/uaccess.h>
 #include <linux/mm_inline.h>
 
+#if defined(CONFIG_KSU_SUSFS_SUS_KSTAT) || defined(CONFIG_KSU_SUSFS_SUS_MAP) || defined(CONFIG_KSU_SUSFS_OPEN_REDIRECT)
+#include <linux/susfs_def.h>
+#endif // #if defined(CONFIG_KSU_SUSFS_SUS_KSTAT) || defined(CONFIG_KSU_SUSFS_SUS_MAP) || defined(CONFIG_KSU_SUSFS_OPEN_REDIRECT)
+
 #include <asm/elf.h>
 #include <asm/tlb.h>
 #include <asm/tlbflush.h>
